Reorganize core build files to be consistent between repos

lwjohnst86 opened the pull request:
@rqding, I am reorganizing many of the files so that our shared common files (`manage.py`, `requirements.txt`, `Dockerfile`, etc) all have consistent and predictable locations. For instance, for us to share the same commands in the `justfile`.

rqding commented:
I am a little bit confused with this story. The manage.py and requirements.txt files belong to the Django and they will not be common files that should be shared because they could contain specific configurations and functions for each repo. I tested the docker build after I pulled the newest code and cleaned up all my local dockers, but it failed to build. So I would not recommend moving those files out of the current registry folder (which is the Django file directory). The Dockerfile is also specific for the Django docker build, so it should still be under the registry folder.

lwjohnst86 commented:
Looking more into the official documentation and thinking more about what we discussed today @rqding, here are a couple of things:

- For Django projects, the `manage.py` by default (via `django startproject`) is in the root directory of the project (https://docs.djangoproject.com/en/4.2/intro/tutorial01/) so where I've moved it now makes sense based on their default structure.
- For Python projects, the `requirements.txt` file should be placed in the root directory (https://www.askpython.com/python/python-requirements-txt), which is where I moved it.
- For Dockerfiles, the default/suggested location is (as you suggested) in the root of the project, so *not* where I moved it. The Docker official docs aren't super clear, but based on their links to "exemplary Dockerfiles" (https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#examples-of-docker-official-images), they are all stored in the root project folder.

So based on that and our discussion, I think if we follow the simple heuristic of "One Git repo = One Django Project (but multiple apps if needed) = One Docker Image", it would make things easier for all of us and the re-organization would make sense in that context since then all build tools are located in the root of the Git project. Do you agree with this?

rqding commented:
@lwjohnst86 Thank you for taking the time to read the documents. 

I think I need to clarify a couple of points.

1. Each of the projects could be one docker framework with multiple images. For example, for the seedcase-registry, this repo already has two images, one Django web image, and one database image. So each repo project could contain multiple directories for each of the docker containers.
2. The Django doc for the manage.py location, I think it means the Django project folder. Let's use the seedcase data node as an example. After the user clone or download the repo, it will add Seedcase folder to the target local location. So under the seedcase directory( seedcase/file.text ) is the Seedcase Project Root directory. Under this folder, there could be multiple directories for each docker container/image. So the Django container/image root folder is Seedcase/Django/ and the manage.py should be at Seedcase/Django/manage.py. There could be other directories for the other containers, such as web_app_angular container, LDAP authentication container, and database container. 
3. You are right about the requirement.py should be in the root directory of any Python project. But because each of the containers has its own building process, it is better to have its own requirements.txt for each of the containers if it needs to install some packages. 


So I will make a project tree for the directories of the seedcase-registry repo, I will add some containers for explain the file structure:

seedcase-registry/
                     Django/
                           base_app/
                              admin.py
                              view.py
                              .....
                           second_app/
                               admin.py
                                view.py
                                .....
                            manage.py
                            requirements.txt
                            Dockerfile
                    Web_app/
                           web_configuration/
                           Dockerfile
                           requirements.txt
                    Database/
                           database_configuration/
                           Dockerfile
                  README.md

lwjohnst86 commented:
Hmmm, I think I understand what you are saying now. Though now it brings up a bigger, fundamental, more philosophical question... Do we follow a mono-repo approach or a multi-repo approach to our all our works (including the website). It sounds like you are advocating for (or have experience in) a monorepo approach to software development, while my experience and personal philosophy is more on the multi-repo approach. I think we should discuss this after you are back from your conference in two weeks. In the meantime, I will write up a pro/con on why I think, for our purposes and context, we should start following a multi-repo approach to all our projects/products.

lwjohnst86 commented:
We will eventually need to reorganize things to be consistent between projects, but we'll hold off until we've dealt with the poly-/mono-repo decision as well as another decision (that I will create an PR for). Converting to draft for now.
